core/chains/evm/client/rpc_client.go

```diff
@@ -0,0 +1,927 @@
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	nodetypes "github.com/smartcontractkit/chainlink/v2/common/chains/client/types"
+	"github.com/smartcontractkit/chainlink/v2/core/assets"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type RPCClient interface {
+	nodetypes.NodeClientAPI[
+		*big.Int,
+		common.Hash,
+		*evmtypes.Head,
+		ethereum.Subscription,
+	]
+	nodetypes.RPCClient[
+		*big.Int,
+		evmtypes.Nonce,
+		common.Address,
+		common.Hash,
+		types.Transaction,
+		common.Hash,
+		types.Log,
+		ethereum.FilterQuery,
+		*evmtypes.Receipt,
+		*assets.Wei,
+		*evmtypes.Head,
+		ethereum.Subscription,
+	]
+}
+
+type rpcClient struct {
+	rpcLog  logger.Logger
+	name    string
+	id      int32
+	chainID *big.Int
+	tier    nodetypes.NodeTier
+
+	ws   rawclient
+	http *rawclient
+
+	stateMu sync.RWMutex // protects state* fields
+
+	// Need to track subscriptions because closing the RPC does not (always?)
+	// close the underlying subscription
+	subs []ethereum.Subscription
+
+	// chStopInFlight can be closed to immediately cancel all in-flight requests on
+	// this rpcClient. Closing and replacing should be serialized through
+	// stateMu since it can happen on state transitions as well as rpcClient Close.
+	chStopInFlight chan struct{}
+}
+
+// NewRPCCLient returns a new *rpcClient as clienttypes.RPCClient
+func NewRPCClient(lggr logger.Logger, wsuri url.URL, httpuri *url.URL, name string, id int32, chainID *big.Int, tier nodetypes.NodeTier) RPCClient {
+	r := new(rpcClient)
+	r.name = name
+	r.id = id
+	r.chainID = chainID
+	r.tier = tier
+	r.ws.uri = wsuri
+	if httpuri != nil {
+		r.http = &rawclient{uri: *httpuri}
+	}
+	r.chStopInFlight = make(chan struct{})
+	lggr = lggr.Named("Client").With(
+		"clientTier", tier.String(),
+		"clientName", name,
+		"client", r.String(),
+		"evmChainID", chainID,
+	)
+	r.rpcLog = lggr.Named("RPC")
+
+	return r
+}
+
+// Not thread-safe, pure dial.
+func (r *rpcClient) Dial(callerCtx context.Context) error {
+	ctx, cancel := r.makeQueryCtx(callerCtx)
+	defer cancel()
+
+	promEVMPoolRPCNodeDials.WithLabelValues(r.chainID.String(), r.name).Inc()
+	lggr := r.rpcLog.With("wsuri", r.ws.uri.Redacted())
+	if r.http != nil {
+		lggr = lggr.With("httpuri", r.http.uri.Redacted())
+	}
+	lggr.Debugw("RPC dial: evmclient.Client#dial")
+
+	wsrpc, err := rpc.DialWebsocket(ctx, r.ws.uri.String(), "")
+	if err != nil {
+		promEVMPoolRPCNodeDialsFailed.WithLabelValues(r.chainID.String(), r.name).Inc()
+		return errors.Wrapf(err, "error while dialing websocket: %v", r.ws.uri.Redacted())
+	}
+
+	r.ws.rpc = wsrpc
+	r.ws.geth = ethclient.NewClient(wsrpc)
+
+	if r.http != nil {
+		if err := r.DialHTTP(); err != nil {
```

Comment:
`SendOnlyNode` requires an HTTP dial instead of WS & HTTP so in order to reuse the same RPC_CLIENT we either have to split `Dial` into `Dial` and `DialHTTP`, or change `Dial` to handle connection based on the `r.tier` value.